compiler/sem/pragmas.nim

```
@@ -1728,60 +1748,171 @@ proc hasPragma*(n: PNode, pragma: TSpecialWord): bool =
+  ## Apply the dynlib pragma from the top of the option stack to `sym`, if
+  ## applicable. The dynlib pragma can be applied if the symbol is marked as
+  ## imported, but no header nor dynlib are specified.
+  let lib = c.optionStack[^1].dynlib
+  if lib != nil and sfImportc in sym.flags and
+     {lfDynamicLib, lfHeader} * sym.loc.flags == {}:
+    incl(sym.loc.flags, lfDynamicLib)
+    addToLib(lib, sym)
+    if sym.loc.r == "":
+      # XXX: this looks like a unnecessary defensive check. If the symbol is
+      #      marked as imported, it already has an external name set
+      sym.loc.r = sym.name.s
+
+proc containsError(n: PNode): bool =
+  for it in n.items:
+    if it.kind == nkError:
+      return true
+
+  result = false
+
+proc pragmaExpr*(c: PContext, n: PNode): PNode =
+  ## Pre-processes the pragma list `n` for the purpose of further processing
+  ## by sem in the context of applying pragmas to an expression.
+  assert n.kind == nkPragma
+  result = copyNode(n)
+  for i in 0..<n.len:
+    prepareSinglePragma(c, n[i], result.sons, exprPragmas, nil)
+
+  for it in result.sons.mitems:
+    if it.kind != nkError:
+      it = processBlockPragma(c, it, whichPragma(it))
+
+    if it.kind == nkError:
+      # wrap the resulting list in an error and abort
+      result = c.config.wrapError(result)
+      break
+
+proc pragmaStmt*(c: PContext, n: PNode, owner: PSym): PNode =
+  ## Analyses and evaluates for their side-effects the pragmas in list `n`.
+  ## `owner` is provided for addition context and is sometimes mutated.
+  assert n.kind == nkPragma
+  result = copyNode(n)
+
+  var hasError = false
+  for i in 0..<n.len:
+    let k = semSinglePragmaInStmt(c, owner, n[i], stmtPragmas, result.sons)
+
+    case k
+    of wPush:
+      hasError = processPush(c, n.info, n.sons.toOpenArray(i+1, n.len-1), result)
+      break
+    of wPop:
+      result.add processPop(c, n[i])
+      # XXX: maybe it'd be better to either report an error if there are items
```

Hmm, on module close if there are unpopped pragmas that's an error? I think that's a good idea. Basically someone forgot to "clean-up" and this would be especially helpful if someone is stuck with push/pop across conditional compilation.

---

The comment refers to all pragmas following a `pop` being silently dropped (e.g. `{.pop, error.}` doesn't abort compilation with a user error). This mirrors the previous behaviour, but I think in the future, the compiler should produce an error (or at least a warning) instead.

Thinking about it more, the three special pragmas should probably also be enforced to also come first (but not as part of this PR).

---

I think the idea at some point was to allow "filtered" pop, so a pop would only pop the items that followed it. This unstructured popping, seems... dangerous.

I agree we should produce an error in the future.